dash/testing/browser.py

```
@@ -0,0 +1,197 @@
+        return (
+            getattr(self, "_get_{}".format(self._browser))()
+            if remote is None
+            else webdriver.Remote(
+                command_executor=remote,
+                desired_capabilities=getattr(
+                    DesiredCapabilities, self._browser.upper()
+                ),
+            )
+        )
+
+    @staticmethod
+    def _get_chrome():
+        options = Options()
+        options.add_argument("--no-sandbox")
+
+        capabilities = DesiredCapabilities.CHROME
+        capabilities["loggingPrefs"] = {"browser": "SEVERE"}
+
+        if "DASH_TEST_CHROMEPATH" in os.environ:
+            options.binary_location = os.environ["DASH_TEST_CHROMEPATH"]
+
+        chrome = webdriver.Chrome(
+            options=options, desired_capabilities=capabilities
+        )
+        chrome.set_window_position(0, 0)
+        return chrome
+
+    @staticmethod
+    def _get_firefox():
+
+        capabilities = DesiredCapabilities.FIREFOX
+        capabilities["loggingPrefs"] = {"browser": "SEVERE"}
+        capabilities["marionette"] = True
+
+        # https://developer.mozilla.org/en-US/docs/Download_Manager_preferences
+        fp = webdriver.FirefoxProfile()
+
+        # this will be useful if we wanna test download csv or other data
+        # files with selenium
+        # TODO this can be fed as a tmpdir fixture from pytest
+        fp.set_preference("browser.download.dir", "/tmp")
+        fp.set_preference("browser.download.folderList", 2)
+        fp.set_preference("browser.download.manager.showWhenStarting", False)
+
+        return webdriver.Firefox(fp, capabilities=capabilities)
+
+    def get_logs(self):
+        """get_logs works only with chrome webdriver"""
+        return (
+            [
+                entry
+                for entry in self.driver.get_log("browser")
+                if entry["timestamp"] > self._last_ts
+            ]
+            if self.driver.name == "chrome"
+            else []
```

empty could mask problems unexpectedly. Would it be better to throw an error if we're not in chrome?

---

the main concern with the throw-an-error approach is that user will be able to choose the browser on the fly, `pytest --webdriver Firefox tests\integration`, and that will raise a lot of errors.

---

True. But we'll also presumably have occasional tests that *require* something to appear in the logs in order to succeed. Maybe returning `[]` is the best we can do for now, but perhaps we should at least print a warning?